### PR TITLE
feat: relation context

### PR DESCRIPTION
**Pull request by LucasFalkowsky**
closes: https://github.com/corona-school/project-user/issues/1030

What was changed:

- userAchievements are now fetched by context
- when adding a userAchievements the context is passed directly or set to null

### CODE REVIEW

File: common/achievement/create.ts [hunk 1]
```diff
@@ -39,7 +47,12 @@ async function createAchievement(templateToCreate: Achievement_template, userId:
...
 }
 
-async function createNextUserAchievement(templatesForGroup: Achievement_template[], nextStepIndex: number, userId: string, context: UserAchievementContext) {
+async function createNextUserAchievement<ID extends ActionID>(
+    templatesForGroup: Achievement_template[],
+    nextStepIndex: number,
+    userId: string,
+    context: SpecificNotificationContext<ID>
+) {
     const nextStepTemplate = templatesForGroup.find((template) => template.groupOrder === nextStepIndex);
```
Comment (LomyW) [inline common/achievement/create.ts:57]: For validation reasons: we could check here whether the correct context for the template exists using the `templateFor` field, as we know here if `templateFor === Match` that a `matchId` / or `relation: "match/1"` must exist, the same applies to subcourse

File: common/notification/actions.ts [hunk 2]
```diff
@@ -638,15 +638,11 @@ const _notificationActions = {
     },
     joined_match_meeting: {
         description: 'User joined a match meeting',
-        sampleContext: {
```
Comment (LomyW) [inline common/notification/actions.ts:641]: I think I would still collect further data in context such as the `matchId / subcourseId  /matchpartner` etc.

File: common/achievement/util.ts [hunk 3]
```diff
@@ -26,10 +26,10 @@ export function getMetricsByAction<ID extends ActionID>(actionId: ID): Metric[]
     return metricsByAction.get(actionId) || [];
 }
 
-type RelationTypes = 'match' | 'subcourse';
+type RelationTypes = 'match' | 'subcourse' | 'achievementName';
```
Comment (LomyW) [inline common/achievement/util.ts:29]: what is the `achievementName` for?

File: common/achievement/create.ts [hunk 4]
```diff
@@ -1,29 +1,37 @@
...
 
-async function doesUserAchievementAlreadyExist(templateId: number, userId: string) {
-    // TODO - check if user achievement exist for one match or one subcourse
+async function doesUserAchievementAlreadyExist<ID extends ActionID>(templateId: number, userId: string, context: SpecificNotificationContext<ID>) {
+    const keys = Object.keys(context);
     const userAchievement = await prisma.user_achievement.findFirst({
         where: {
             templateId,
             userId,
+            AND: keys.map((key) => {
```
Comment (LomyW) [inline common/achievement/create.ts:12]: So with this check we can guarantee that the user_achievement for a user exists for a specific match / subcourse?

**Review by LomyW — CHANGES_REQUESTED** [review]
Nice ✨

File: common/notification/actions.ts (hunk 2 above)
Comment (LucasFalkowsky) [reply common/notification/actions.ts:641]: and we do in the mutations file. As I see it, this is just a sample and not responsible for fetching all necessary data, as this is done by the mutation. Am I wrong?

File: common/achievement/create.ts (hunk 1 above)
Comment (LucasFalkowsky) [reply common/achievement/create.ts:58]: but we do pass in the context from the previous userAchievement. Isn't that what we want here? like:
{ matchId: match/1 } from previous userAchievement gets also written in this achievements context.

File: common/notification/actions.ts (hunk 2 above)
Comment (LucasFalkowsky) [reply common/notification/actions.ts:641]: I talked to our beloved boss and he explained to me that the sample context is just for visualization and is never used. I adjusted my code accordingly.

Comment (Jonasdoubleyou) [reply common/notification/actions.ts]: "and he explained to me that the sample context is just for visualization and is never used" - This is not correct. The sample context is used to validate Notifications created for a certain action to ensure all templating done is correct. Also it is used to infer the type of `SpecificNotificationContext`. So please maintain this.

File: common/achievement/util.ts [hunk 5]
```diff
@@ -39,9 +39,9 @@ export function getRelationTypeAndId(relation: string): [relationType: RelationT
...
 }
 
-export async function getRelationContext(relation: string): Promise<RelationContextType> {
+export async function getAchievementContext(relation: string): Promise<AchievementContextType> {
     const [type, id] = getRelationTypeAndId(relation);
-    const relationContext: RelationContextType = {
+    const achievementContext: AchievementContextType = {
         match:
             type === 'match'
                 ? await prisma.match.findFirst({ where: { id }, select: { id: true, lecture: { select: { start: true, duration: true } } } })[0]
```
Comment (LomyW) [inline common/achievement/util.ts:47]: Is this working `[0]`?

File: common/achievement/types.ts [hunk 6]
```diff
@@ -111,7 +111,7 @@ export type EvaluationResult = {
     resultObject: Record<string, string | number | boolean>;
 };
 
-export type RelationContextType = {
+export type AchievementContextType = {
     match?: {
         id: number;
         lecture: {
```
Comment (LomyW) [inline common/achievement/types.ts:117]: Here I think we need an array of lectures or? So that we can create a bucket for each match appointment or?

**Review by LomyW — CHANGES_REQUESTED** [review]

**Pull request merged**